### PR TITLE
Http index integration

### PR DESCRIPTION
**Pull request by ccmoloney**
Interim merge.  Updated to handle WPS-like parameters, ie: layerName,  subset & outputFile.
Added code to aggregation worker to source a file list from geoserver + run an aggregation.
Various 'fixes' and re-organisations.

### CODE REVIEW

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 1]
```diff
@@ -69,33 +80,47 @@ public void run(String... args) {
...
+
+            //  Check parameters:
+            //      Expecting the following params -
+            //        - layerName
+            //        - subset
+            //        - outputFile
+
+            String layerName = commandLine.getArgs()[0];
+            String subset = commandLine.getArgs()[1];
+            String destinationFile = commandLine.getArgs()[2];
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Probably can determine the output file location here using config and the job id at some point.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    public URIList getUriList(String profile, String timeField, String urlField, SubsetParameters subset) throws AggregationException {
+        String timeCoverageStart = subset.get("TIME").start;
+        String timeCoverageEnd = subset.get("TIME").end;
+
+        URIList uriList = new URIList();
+
+        try
+        {
+            String downloadUrl = String.format("%s/wfs", geoserver);
+            String cqlFilter = String.format("%s >= %s AND %s <= %s",
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Need to handle optional time range at some point

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                        uriList.add(new URI(lineParts[2]));
+                    } else {
+                        logger.debug("CSV headers = " + line);
+                    }
+                    i++;
+                }
+            }
+            finally
+            {
+                if(inputStream != null) { inputStream.close(); }
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Can we use try with resources block instead? - its much cleaner

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(HttpIndexReader.class);
+
+    protected String geoserver = null;
+
+    public HttpIndexReader(String geoserver) {
+        this.geoserver = geoserver;
+    }
+
+    @Override
+    public URIList getUriList(String profile, String timeField, String urlField, SubsetParameters subset) throws AggregationException {
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: perhaps rename profile as layer - its what this actually is - profile refers to a very old gogoduck config mechanism.

Comment (jonescc) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: also probably just need the time parameters here.

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 5]
```diff
@@ -69,33 +80,47 @@ public void run(String... args) {
...
+                {
+                    logger.debug(" - " + file.toString());
+                }
             }
 
-            String destination = line.getArgs()[1];
-            S3URI s3URI = new S3URI(destination);
 
-            Set<DownloadRequest> downloads = new LinkedHashSet<>();
+            Set<DownloadRequest> downloads = indexReader.getDownloadRequestList(layerName, "time", "file_url", subsetParams);
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Don't need getUriList above anymore?

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        catch (Exception e) {
+            logger.error("We could not obtain list of URLs, does the collection still exist?");
+            throw new AggregationException(String.format("Could not obtain list of URLs: '%s'", e.getMessage()));
+        }
+
+        return uriList;
+    }
+
+
+    public Set<DownloadRequest> getDownloadRequestList(String profile, String timeField, String urlField, SubsetParameters subset) throws AggregationException {
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Comments as above for getUriList.  Code is pretty similar - but suspect we are getting rid of the one above?

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                if(bufferedStream != null) { bufferedStream.close(); }
+
+                if(dataInputStream != null) { dataInputStream.close(); }
+
+                if(conn != null) { conn.disconnect(); }
+            }
+        }
+        catch (Exception e) {
+            logger.error("We could not obtain list of URLs, does the collection still exist?");
+            throw new AggregationException(String.format("Could not obtain list of URLs: '%s'", e.getMessage()));
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: mesage needs to end up in the status document at some point (failed status)

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            HttpURLConnection conn = null;
+            InputStream inputStream = null;
+            BufferedInputStream bufferedStream = null;
+            DataInputStream dataInputStream = null;
+
+            try
+            {
+                //  Make HTTP request to geoserver
+                conn = (HttpURLConnection) url.openConnection();
+                conn.setRequestMethod("GET");
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: GET here but talking about post elsewhere and - writing to request stream.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/SubsetParameters.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            put(subsetParts[0], new SubsetParameter(subsetParts[1], subsetParts[2]));
+        }
+    }
+
+    // Simple copy ctor
+    public SubsetParameters(SubsetParameters sp) {
+        super(sp);
+    }
+
+    public synchronized List<String> getNcksParameters() {
```
Comment (jonescc) [inline aggregation-worker/src/main/java/au/org/emii/geoserver/client/SubsetParameters.java]: This isn't relevant anymore - and not used above

File: request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java [hunk 10]
```diff
@@ -35,90 +28,30 @@ public AwsApiResponse handleRequest(AwsApiRequest request, Context context) {
...
             //  Execute the request
             response = handler.handleRequest(request, config);
 
         }
-        catch(IOException ioex)
+        catch(Exception ex)
         {
             //  Bad stuff happened
-            LOGGER.log("Exception running WPS Lambda function: " + ioex.getMessage());
+            LOGGER.log("Exception running WPS Lambda function: " + ex.getMessage());
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java]: Do we need to setup an error reponse here?

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/S3StatusUpdater.java [hunk 11]
```diff
@@ -25,10 +26,26 @@ public S3StatusUpdater(String s3bucket, String uniqueId) {
     }
 
     public void updateStatus (String statusDocument) throws UnsupportedEncodingException {
-        AmazonS3 cl = AmazonS3ClientBuilder.defaultClient();
-        cl.putObject(
-                new PutObjectRequest(s3bucket, uniqueId + "/" + statusFilename, new StringInputStream(statusDocument), new ObjectMetadata())
-                        .withCannedAcl(CannedAccessControlList.PublicRead));
+        System.out.println("Updating status in S3 : " + statusDocument);
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/S3StatusUpdater.java]: perhaps use logger at some point so we can log different amounts of info in dev to prod

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java [hunk 12]
```diff
@@ -82,11 +82,11 @@ public String execute(Properties config) {
...
         LOGGER.debug("Job submitted.  Job ID : " + jobId);
 
-        String statusLocation = statusLocationBase + "prototypeId" + "/" + statusFileName;
-        ExecuteStatusBuilder statusBuilder = new ExecuteStatusBuilder(statusLocation,jobId);
+        String statusLocation = "https://s3.amazonaws.com/" + statusS3BucketName + "/" + jobId + "/" + statusFileName;
+        ExecuteStatusBuilder statusBuilder = new ExecuteStatusBuilder(statusLocation, jobId);
         String statusDocument = statusBuilder.createResponseDocument(EnumStatus.ACCEPTED);
 
-        S3StatusUpdater statusUpdater = new S3StatusUpdater(statusLocationBase, "prototypeId");
+        S3StatusUpdater statusUpdater = new S3StatusUpdater(statusS3BucketName, jobId);
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java]: Does this have code to generate the status location as well?  would it better to consolidate this into one location?

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/WpsConfig.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        S3ObjectInputStream contentStream = configFile.getObjectContent();
+        Properties config = new Properties();
+        try {
+            //  Load configuration from S3 file
+            config.load(contentStream);
+        }
+        catch(Exception ex)
+        {
+            ex.printStackTrace();
+            System.out.println("Unable to load configuration : " + ex.getMessage());
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/WpsConfig.java]: Should this stop further processing?

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 1 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Yep - is the target location for the aggregated file the same location as the status file?  Or will we have another base location for the aggregated file?

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 5 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: No - probably not needed.

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java [hunk 14]
```diff
@@ -69,33 +80,47 @@ public void run(String... args) {
...
+            S3URI s3URI = new S3URI(destinationFile);
 
-            for (String inputFile: inputFiles.subList(1, inputFiles.size()-1)) {
-                if (inputFile.trim().isEmpty()) continue;
-                URL fileUrl = new URL("http://s3-ap-southeast-2.amazonaws.com/imos-data/" + inputFile.split(",")[1]);
-                long size = Long.parseLong(inputFile.split(",")[2]);
-                downloads.add(new DownloadRequest(fileUrl, size));
-            }
+            String overridesArg = commandLine.getOptionValue("c");
+            String bboxArg = commandLine.getOptionValue("b");
```
[the remark this answers is not in the document]
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: OK - will have a look at how GoGoDuck does it.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java (hunk 4 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Time params are part of the SubsetParameters object.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java (hunk 2 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: As in cater for the case where no time params are passed?

File: aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java (hunk 1 above)
Comment (jonescc) [reply aggregation-worker/src/main/java/au/org/emii/aggregationworker/AggregationRunner.java]: Can be in the same folder - but the status file is separate to the output file

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java (hunk 2 above)
Comment (jonescc) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Yes - time is an optional subset parameter - some datasets don't include time and a time range wont be passed for them.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java (hunk 3 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Sure.

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/SubsetParameters.java (hunk 9 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/SubsetParameters.java]: Will remove.

File: request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java (hunk 10 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java]: Yep - will add TODO

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/S3StatusUpdater.java (hunk 11 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/S3StatusUpdater.java]: I knew you'd say that  ;)

File: aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java (hunk 6 above)
Comment (ccmoloney) [reply aggregation-worker/src/main/java/au/org/emii/geoserver/client/HttpIndexReader.java]: Removed getUriList method

File: request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java (hunk 10 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/lambda/WpsLambdaRequestHandler.java]: Added TODO

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/WpsConfig.java (hunk 13 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/WpsConfig.java]: Added TODOs to resolve what happens if config can't be loaded.

**Pull request merged**